gemini/reasoning-engine/tutorial_graph_rag.ipynb

```diff
@@ -0,0 +1,1707 @@
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "O-lCGcLP1loV",
+    "outputId": "bfd3647c-8073-48a9-f771-0ebeaf445b3a",
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Lawrence Edward Page WORKED_AT Alphabet Inc.\n",
+      "Lawrence Edward Page WORKED_AT Google\n",
+      "Lawrence Edward Page INVESTMENT Kitty Hawk\n",
+      "Lawrence Edward Page INVESTMENT Opener\n",
+      "Lawrence Edward Page INFLUENCED_BY Maria Montessori\n"
+     ]
+    }
+   ],
+   "source": [
+    "from google.cloud import spanner\n",
+    "\n",
+    "spanner_client = spanner.Client(GCP_PROJECT_ID)\n",
+    "instance = spanner_client.instance(SPANNER_INSTANCE_ID)\n",
+    "database = instance.database(SPANNER_DATABASE_ID)\n",
+    "\n",
+    "\n",
+    "tables = [\"NATIONALITY\", \"WORKED_AT\", \"INVESTMENT\", \"INFLUENCED_BY\"]\n",
+    "a = []\n",
+    "for table in tables:\n",
+    "    with database.snapshot() as snapshot:\n",
+    "        if table in (\"INFLUENCED_BY\", \"LOCATED_IN\"):\n",
+    "            column_name = \"P1_Name\"\n",
+    "        else:\n",
+    "            column_name = \"P_Name\"\n",
+    "\n",
+    "        results = snapshot.execute_sql(\n",
+    "            f\"\"\"\n",
+    "            SELECT {column_name} AS Name, *\n",
+    "            FROM {table}\n",
+    "            WHERE {column_name} = '{kgnodename}'\n",
+    "            \"\"\"\n",
+    "        )\n",
+    "        for row in results:\n",
+    "            # Dynamically determine the index of the relevant column\n",
+    "            name_index = row.index(kgnodename)\n",
+    "            # Construct the output string using f-string\n",
+    "            output_string = f\"{row[name_index]} {table} {row[2]}\"\n",
+    "            print(output_string)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "gTdtp14I7P5g"
+   },
+   "source": [
+    "### Langchain Custom Retriever and Chain\n",
+    "\n",
+    "We will create a custom Langchain retriever that will encapsulate code blocks above and chain together to make a functional Q&A bot."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {
+    "id": "qzOZ5NUv55vU",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "from typing import List\n",
+    "\n",
+    "from langchain_core.callbacks import CallbackManagerForRetrieverRun\n",
+    "from langchain_core.documents import Document\n",
+    "from langchain_core.retrievers import BaseRetriever\n",
+    "from vertexai.language_models import TextEmbeddingInput, TextEmbeddingModel\n",
+    "from langchain_google_vertexai import VertexAIEmbeddings\n",
+    "from google.cloud import spanner\n",
+    "\n",
+    "class SpannerGraphRagRetriever(BaseRetriever):\n",
+    "  gcp_project_id:str\n",
+    "  spanner_instance_id:str\n",
+    "  spanner_database_id:str\n",
+    "  embedding_model_name:str\n",
```

Comment:
Do you need to build a custom retriever for this? There should already be a Spanner Retriever in LangChain.

Also, would it be possible to just define these methods once in the Chain instead of running them multiple times in the notebook? It seems like there's lots of duplication.